### ISSUE TITLE
Firebase Cloud Function Socket Hangup Error & No Email Sent

### ISSUE THREAD

**Issue by Richard4097** [sendgrid/sendgrid-nodejs#1268]
### Issue Summary
Hello, I am trying to use a firebase functions PubSub to look through a collection on my Firestore database and then send emails to people based when they were created, using different dynamic template. EX: on Day 2 we send an email on Tips & Tricks for the site, on day 3 we send an email asking for feedback. When doing this in small numbers (less than 100) it works fine, but when those numbers start to go higher, 100-200, first the function finishes with status "ok", but then no emails are sent and I receive a socket hangup error. When commenting out the line "return sgMail.send(msg)" the function then executes fine.

### Code Snippet
```node
try {
        var users = await db.collection("users").where("live", "==", true).get()
        users.forEach(doc => {
            userData = doc.data()
            userData.uid = doc.id
            if(userData.createdAt && userData.lastActive){
                var createdAt = userData.createdAt.seconds;
                var lastActive = userData.lastActive.seconds;
                var currentTime = Date.now() / 1000;
                var difference = (currentTime - createdAt) / 24 / 60 / 60;

                if (difference > 1 && difference <= 2) {
                        const msg = {
                            to: email,
                            from: {
                                email: 'ouremail',
                                name: 'ourname'
                            },
                            templateId:'ourtemplateID',
                            dynamicTemplateData: {
                                first_name: firstName,
                            },
                            asm: {
                                groupId: 15816,
                            }
                        };
                        console.log('Sent Email About Promotion')
                        return sgMail.send(msg);
                }

                if(difference > 2 && difference <= 3){
                         const msg = {
                                to: email,
                                from: {
                                    email: ouremail',
                                    name: 'ourname'
                                },
                                templateId:'ourtemplateID',
                                dynamicTemplateData: {
                                    first_name: firstName
                                },
                                asm: {
                                    groupId: 15816,
                                }
                            };
                       console.log('Sent Tips Email Seeker')
                       return sgMail.send(msg);
                }
        })
        return users
} catch (err) {
       console.log(err)
       return 
}
```
I apologize for the worlds worst indentation ^

### Exception/Log

8:26:07.494 AM | warning | Process exited with code 16    
      at process.on.code (/layers/google.nodejs.functions-framework/functions-framework/node_modules/@google-cloud/functions-framework/build/src/invoker.js:275:22)     
     at process.emit (events.js:198:13)     at process.EventEmitter.emit (domain.js:448:20)
     at process.exit (internal/process/per_thread.js:168:15)
     at Object.sendCrashResponse (/layers/google.nodejs.functions-framework/functions-framework/node_modules/@google-cloud/functions-framework/build/src/logger.js:37:9)
     at process.on.err (/layers/google.nodejs.functions-framework/functions-framework/node_modules/@google-cloud/functions-framework/build/src/invoker.js:271:22)
     at process.emit (events.js:198:13)
     at process.EventEmitter.emit (domain.js:448:20)
     at emitPromiseRejectionWarnings (internal/process/promises.js:140:18)
     at process._tickCallback (internal/process/next_tick.js:69:34) | more_vert

8:26:06.488 AM | warning | Error: socket hang up at createHangUpError (_http_client.js:332:15) at TLSSocket.socketOnEnd (_http_client.js:435:23) at TLSSocket.emit (events.js:203:15) at TLSSocket.EventEmitter.emit (domain.js:466:23) at endReadableNT (_stream_readable.js:1145:12) at process._tickCallback (internal/process/next_tick.js:63:19)

8:26:06.387 AM | Unhandled rejection

8:25:04.210 AM | Function execution took 2105 ms, finished with status: 'ok'

### Technical details:
"@sendgrid/mail": "^7.2.0",

**Comment by shwetha-manvinkurke** [comment]
Hi @Richard4097 [This](https://stackoverflow.com/questions/16995184/nodejs-what-does-socket-hang-up-actually-mean) might help? I suggest you reach out to https://support.sendgrid.com if you are still facing the issue.

**Comment by Richard4097** [comment]
Thanks for the help, that post lead me to another that helped me better understand how to debug the problem! Understanding the sgMail.Send() returns a promise, that you can chain and catch errors from was something that I was missing.

**Comment by thinkingserious** [comment]
Hello @Richard4097,

Can you please take a look at [this solution](https://github.com/sendgrid/sendgrid-nodejs/issues/891#issuecomment-819766766) and let us know if that fixes your issue.

With best regards,

Elmer

**Comment by douglasrcjames** [comment]
@Richard4097 getting this exact issue, any idea how exactly you resolved this "socket hangup" error?

**Comment by Richard4097** [comment]
To this day I'm still not exactly sure what the issue was, but I know that a combination of 3 things helped me resolve the problem. 

1. I handled the sgmail.send as a promise, something like... 

`return sgMail.send(msg).then(() => console.log('Sent Email' + firstName)).catch(error => 
console.log(error));`

Instead of just sgmail.send(msg)

Down the line I also switched over to a try, catch with awaits instead of promise chaining.

2. I checked through my database and cleaned out some really old data that didnt have the correct formatting/fields. This I expected to do it on its own, but for some reason did not. 

3. I reduced the amount of sends in the loop by breaking my function apart into two similar functions. I was unhappy with this because it seemed like I was creating redundant code, and I was no where near the limits for the api, but again, it got the job done, so...

I never got around to figuring out the exact root of the problem and coming up with a more precise/elegant solution, as it was a smaller concern in my larger build, and the problem was ultimately solved, but this combination of things did it for me. Hope this helps, wish I could be more help.
